beacon-chain/sync/validate_aggregate_proof.go

```diff
@@ -0,0 +1,127 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/blocks"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// validateAggregateAndProof verifies the aggregated signature and the selection proof is valid before forwarding to the
+// network and downstream services.
+func (r *RegularSync) validateAggregateAndProof(ctx context.Context, msg proto.Message, p p2p.Broadcaster, fromSelf bool) (bool, error) {
+	// To process the following it requires the recent blocks to be present in the database, so we'll skip
+	// validating or processing aggregated attestations until fully synced.
+	if r.initialSync.Syncing() {
+		return false, nil
+	}
+
+	m := msg.(*pb.AggregateAndProof)
+	attSlot := m.Aggregate.Data.Slot
+
+	// Verify aggregate attestation has not already been seen via aggregate gossip, within a block, or through the creation locally.
+	// TODO(3835): Blocked by operation pool redesign
+
+	// Verify the block being voted for passes validation. The block should have passed validation if it's in the DB.
+	if !r.db.HasBlock(ctx, bytesutil.ToBytes32(m.Aggregate.Data.BeaconBlockRoot)) {
+		return false, errPointsToBlockNotInDatabase
+	}
+
+	s, err := r.chain.HeadState(ctx)
+	if err != nil {
+		return false, err
+	}
+	if attSlot > s.Slot {
+		s, err = state.ProcessSlots(ctx, s, attSlot)
```

Comment:
This is potentially a griefing factor / attack vector. Can you do this last, after passing all other validations?

Reply:
No because we'll need the latest state to verify aggregation is within the correct committee.

I will add a check to discard attestation from the future slot

Reply:
Nvm, there's an easier fix. Just had to move the attestation slot check before this